### PR TITLE
convert codebase to ESM & add build `watch` support

### PR DESCRIPTION
**Pull request by o-az**
Convert to ESM & add watch support for build.

Converting to ESM required

- updating a number of `tsconfig` settings, including `moduleResolution` to be `Bundler`,
- postfixing local import with `.js` (in spite of [this](https://www.typescriptlang.org/tsconfig#moduleResolution)),
- replacing `__dirname` with ESM-safe option,
- leveraging `createRequire` for some imports,
- updating `type` to `module`,
- updating some dependencies to latest (each upgraded dependency had to be upgraded for this to work),
- small change to logic in `core/src/build/handlers.ts#buildRawHandlerFunctions`,
- small change to logic in `core/src/config/config.ts#buildConfig`,
- small tweak to `vitest` config needed for 2 tests (ens & art-gobblers) specifically,
- [import assertions](https://v8.dev/features/import-assertions) for json imports,
- switching to default export for `pg` (and updating the dep),
- small updates to types here and there,
- other things I can't remember atm :>

### CODE REVIEW

**Comment by o-az** [comment]
@0xOlias all tests pass except for ens and art-gobblers in core

File: packages/core/src/_test/ens/app/src/index.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import { ponder } from "@/generated";
+import { ponder } from "../generated/index.js";
```
Comment (0xOlias) [inline packages/core/src/_test/ens/app/src/index.ts:1]: Curious why this was necessary. Does the alias still work in the example projects?

File: packages/core/src/build/handlers.ts [hunk 2]
```diff
@@ -1,15 +1,19 @@
...
-import type { Log } from "@/types/log";
-import type { Transaction } from "@/types/transaction";
+import type { LogEventMetadata, LogFilter } from "@/config/logFilters.js";
+import type { Options } from "@/config/options.js";
+import type { Block } from "@/types/block.js";
+import type { Log } from "@/types/log.js";
+import type { Transaction } from "@/types/transaction.js";
+import { isRejected } from "@/utils/promise.js";
+
+const require = createRequire(import.meta.url);
```
Comment (0xOlias) [inline packages/core/src/build/handlers.ts:16]: Did you test that hot reloading still works using createRequire with the require cache invalidation trick?

File: packages/core/src/build/handlers.ts [hunk 3]
```diff
@@ -148,23 +152,18 @@ export const buildRawHandlerFunctions = async ({
...
-      } catch (err) {
-        return err as Error;
-      }
-    })
-    .filter((err): err is Error => err !== undefined);
+  const importUserFiles = await Promise.allSettled(
+    outUserFilenames.map((filename) => import(filename))
+  );
+
+  const rejecteds = importUserFiles.filter(isRejected);
```
Comment (0xOlias) [inline packages/core/src/build/handlers.ts:159]: This change to allSettled makes sense, though I'd probably call this `rejectedImports`.

**Review by 0xOlias — COMMENTED** [review]
Couple small questions. Seems like most of the diff is updating imports to use the `.js` prefix. Any other areas you were unsure about / I should take a close at?

File: packages/core/src/build/handlers.ts (hunk 2 above)
Comment (o-az) [reply packages/core/src/build/handlers.ts:16]: yup first thing. Sadly doesn't work

File: packages/core/src/build/handlers.ts (hunk 3 above)
Comment (o-az) [reply packages/core/src/build/handlers.ts]: done

File: packages/core/src/_test/ens/app/src/index.ts (hunk 1 above)
Comment (o-az) [reply packages/core/src/_test/ens/app/src/index.ts:1]: Actually not necessary after I added the vitest resolve. So reverting

Comment (o-az) [reply packages/core/src/_test/ens/app/src/index.ts]: done

File: packages/core/src/build/handlers.ts (hunk 2 above)
Comment (o-az) [reply packages/core/src/build/handlers.ts:16]: m8 be relevant https://github.com/evanw/esbuild/issues/1828

**Comment by o-az** [comment]
> Couple small questions. Seems like most of the diff is updating imports to use the `.js` prefix. Any other areas you were unsure about / I should take a close at?

If you could search globally for "await import" and review those files that would be great.

Besides that majority of it is adding .js

**Pull request closed without merging**